api/group_creation/group_manager.py

```diff
@@ -0,0 +1,150 @@
+        def is_acceptable(self) -> bool:
+            # Checks group size is appropriate for age range
+            restrictions = GroupManager._get_allowed_group_size(self.age_range)
+            students_num = len(self.students)
+            if restrictions.min_number > students_num:
+                return False
+            if restrictions.max_number < students_num:
+                raise ValueError(
+                    f"Group candidate must respect upper group boundaries, "
+                    f"but {students_num} > {restrictions.max_number}"
+                )
+            return True
+
+    @staticmethod
+    def get_available_teachers() -> Iterable[Teacher]:
+        return Teacher.objects.filter(
+            status__in=(
+                TeacherStatus.AWAITING_OFFER,
+                TeacherStatus.TEACHING_ACCEPTING_MORE,
+            )
+        )
+
+    @staticmethod
+    def create_group(teacher_id: int) -> Group | None:
+        for group_candidate in GroupManager._iterate_group_candidates(teacher_id):
+            if group_candidate.is_acceptable():
+                group = Group(
+                    language_and_level=group_candidate.language_and_level,
+                    communication_language_mode=group_candidate.communication_language_mode,
+                    lesson_duration_in_minutes=DEFAULT_LESSON_DURATION_MIN,
+                    monday=group_candidate.monday,
+                    tuesday=group_candidate.tuesday,
+                    wednesday=group_candidate.wednesday,
+                    thursday=group_candidate.thursday,
+                    friday=group_candidate.friday,
+                    saturday=group_candidate.saturday,
+                    sunday=group_candidate.sunday,
+                )
+
+                group_creation_timestamp = timezone.now()
+                StatusSetter.set_status(
+                    obj=group, status=GroupStatus.PENDING, status_since=group_creation_timestamp
+                )
+                group.save()
+                group.teachers.add(group_candidate.teacher)
+                group.students.set(group_candidate.students)
+
+                # QQ: do we update teacher/student status automatically,
+                # or wait for coordinator action?
+                next_teacher_status = (
+                    TeacherStatus.TEACHING_ANOTHER_GROUP_OFFERED
+                    if group_candidate.teacher.status == TeacherStatus.TEACHING_ACCEPTING_MORE
+                    else TeacherStatus.GROUP_OFFERED
+                )
+                StatusSetter.set_status(
+                    obj=group_candidate.teacher,
+                    status=next_teacher_status,
+                    status_since=group_creation_timestamp,
+                )
+                GroupManager._create_log_events(group)
+                return group
+        # No suitable groups found
+        return None
+
+    @staticmethod
+    def _get_allowed_group_size(age_range: AgeRange) -> GroupSizeRestriction:
```

Comment:
Note: I don't like this method, hope to come up with more readable code eventually

Reply:
I'm not commenting on this bit, then :)

Reply:
If you have any specific suggestions, feel free to let it out! :)

Reply:
> If you have any specific suggestions, feel free to let it out! :)

Nah, I was being lazy and skipped the whole bit 😜